ddtrace/contrib/datadog_lambda/patch.py

```diff
@@ -0,0 +1,148 @@
+from importlib import import_module
+import os
+import signal
+
+from ddtrace import tracer
+from ddtrace.constants import ERROR_MSG
+from ddtrace.constants import ERROR_TYPE
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.utils import get_argument_value
+from ddtrace.internal.wrapping import unwrap
+from ddtrace.internal.wrapping import wrap
+
+
+log = get_logger(__name__)
+
+
+def crash_flush(_, __):
+    """
+    Tags the current root span with an Impending Timeout error and
+    finishes all remaining spans.
+
+    Shutdowns the tracer at the end.
+    """
+
+    root_span = tracer.current_root_span()
+    root_span.error = 1
+    root_span.set_tag_str(ERROR_MSG, "Datadog detected an Impending Timeout")
+    root_span.set_tag_str(ERROR_TYPE, "Impending Timeout")
+
+    current_span = tracer.current_span()
+    while current_span:
+        current_span.finish()
+        current_span = current_span._parent
+
+
+def check_timeout(context):
+    """
+    Creates a timeout to detect when an AWS Lambda handler's remaining
+    time is about to end.
+
+    Crashes flushes when the signal is activated
+    """
+    signal.signal(signal.SIGALRM, crash_flush)
+    remaining_time_in_millis = context.get_remaining_time_in_millis()
+    apm_flush_deadline = int(os.environ.get("DD_APM_FLUSH_DEADLINE", 0))
+
+    if apm_flush_deadline and apm_flush_deadline <= remaining_time_in_millis:
+        if apm_flush_deadline < 200:
```

Comment:
this codepath gets followed in the default case, should we just set the default to 200?

Reply:
We can't have it as a default, because we expect the user to set their own deadline as desired, which would take priority over `remaining_time_in_millis`. Unless it's set as not 0, we should change it to 200 to still have 100ms to flush.

Reply:
ah I see. I missed the first part condition check `if apm_flush_deadline and`, could we make this more explicit `if apm_flush_deadline > 0` for readability.

Also a comment would be good for why 200ms is required as the default. Can we also comment why in the warning log we use 200ms to be a bit more user friendly.

Reply:
Yup, on it